packages/services/src/contents/index.ts

```diff
@@ -1571,3 +1575,259 @@ namespace Private {
+    }
+    const data = await response.json();
+    validate.validateContentsModel(data);
+    this.fileChanged.emit({
+      type: 'save',
+      oldValue: null,
+      newValue: data
+    });
+    return data;
+  }
+
+  /**
+   * A signal emitted when a file operation takes place.
+   */
+  get fileChanged(): Signal<this, Contents.IChangedArgs> {
+    return this.drive.fileChanged as unknown as Signal<
+      this,
+      Contents.IChangedArgs
+    >;
+  }
+
+  /**
+   * Get a REST url for a file given a path.
+   */
+  private _getUrl(...args: string[]): string {
+    const parts = args.map(path => URLExt.encodeParts(path));
+    const baseUrl = this.drive.serverSettings.baseUrl;
+    return URLExt.join(baseUrl, this._apiEndpoint, ...parts);
+  }
+
+  private _apiEndpoint: string;
+  private _drive: Contents.IDrive;
+}
+
+/**
+ * the content provider registry interface.
+ */
+export interface IContentProviderRegistry {
+  /**
+   * the drive of the content provider registry.
+   */
+  drive: Contents.IDrive | undefined;
+
+  /**
+   * Add a content provider into the registry.
+   *
+   * @param provider - The content provider to register.
+   */
+  register(provider: IContentProvider): void;
+
+  /**
+   * Get a content provider for the given file path.
+   *
+   * @param filePath - The file path for which to get a content provider.
+   */
+  getProvider(filePath: string): IContentProvider | undefined;
+}
+
+/**
+ * The content provider interface.
+ */
+export interface IContentProvider {
+  /**
+   * The drive of the content provider.
+   */
+  drive: Contents.IDrive;
+
+  /**
+   * An optional shared model factory instance for the content provider.
+   */
+  readonly sharedModelFactory?: Contents.ISharedFactory;
+
+  /**
+   * The file extensions that this provider handles.
+   */
+  readonly extensions: ContentProviderExtensions;
+
+  /**
+   * Get the file content.
+   *
+   * @param localPath - The path to the file.
+   *
+   * @param options - The optional options for fetching the content.
+   *
+   * @returns A promise which resolves with the file content model.
+   */
+  get(
+    localPath: string,
+    options?: Contents.IFetchOptions
+  ): Promise<Contents.IModel>;
+
+  /**
+   * Save a file.
+   *
+   * @param localPath - The desired file path.
+   *
+   * @param options - Optional overrides to the model.
+   *
+   * @returns A promise which resolves with the file content model when the file is saved.
+   */
+  save(
+    localPath: string,
+    options: Partial<Contents.IModel>
+  ): Promise<Contents.IModel>;
+
+  /**
+   * A signal emitted when a file operation takes place.
+   */
+  readonly fileChanged: ISignal<this, Contents.IChangedArgs>;
+}
+
+/**
+ * The content provider extensions interface, consisting of the regural expression matching the file
+ * extension and the score indicating how good the provider is at handling this type of content.
+ */
+export type ContentProviderExtensions = { re: string; score: number }[];
```

Comment:
I would suggest to define `IContentProviderExtension` interface and use it like `IContentProviderExtension[]`. Using interfaces is better for extension authors and for debugging in TS over types.